eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs

```diff
@@ -0,0 +1,186 @@
+  TRC (..),
+  TransitionRule,
+  judgmentContext,
+  trans,
+ )
+import Data.Default.Class (Default)
+import qualified Data.Map.Strict as Map
+
+data ConwayEpochEvent era
+  = PoolReapEvent (Event (EraRule "POOLREAP" era))
+  | SnapEvent (Event (EraRule "SNAP" era))
+
+instance
+  ( EraTxOut era
+  , Embed (EraRule "SNAP" era) (ConwayEPOCH era)
+  , Environment (EraRule "SNAP" era) ~ LedgerState era
+  , State (EraRule "SNAP" era) ~ SnapShots (EraCrypto era)
+  , Signal (EraRule "SNAP" era) ~ ()
+  , Embed (EraRule "POOLREAP" era) (ConwayEPOCH era)
+  , Environment (EraRule "POOLREAP" era) ~ PParams era
+  , State (EraRule "POOLREAP" era) ~ ShelleyPoolreapState era
+  , Signal (EraRule "POOLREAP" era) ~ EpochNo
+  , Default (PParams era)
+  , Eq (UpecPredFailure era)
+  , Show (UpecPredFailure era)
+  , Default (PPUPState era)
+  , EraTallyState era
+  ) =>
+  STS (ConwayEPOCH era)
+  where
+  type State (ConwayEPOCH era) = EpochState era
+  type Signal (ConwayEPOCH era) = EpochNo
+  type Environment (ConwayEPOCH era) = ()
+  type BaseM (ConwayEPOCH era) = ShelleyBase
+  type PredicateFailure (ConwayEPOCH era) = ShelleyEpochPredFailure era
+  type Event (ConwayEPOCH era) = ConwayEpochEvent era
+  transitionRules = [epochTransition]
+
+epochTransition ::
+  forall era.
+  ( Embed (EraRule "SNAP" era) (ConwayEPOCH era)
+  , Environment (EraRule "SNAP" era) ~ LedgerState era
+  , State (EraRule "SNAP" era) ~ SnapShots (EraCrypto era)
+  , Signal (EraRule "SNAP" era) ~ ()
+  , Embed (EraRule "POOLREAP" era) (ConwayEPOCH era)
+  , Environment (EraRule "POOLREAP" era) ~ PParams era
+  , State (EraRule "POOLREAP" era) ~ ShelleyPoolreapState era
+  , Signal (EraRule "POOLREAP" era) ~ EpochNo
+  ) =>
+  TransitionRule (ConwayEPOCH era)
+epochTransition = do
+  TRC
+    ( _
+      , EpochState
+          { esAccountState = acnt
+          , esSnapshots = ss
+          , esLState = ls
+          , esPrevPp = pr
+          , esPp = pp
+          , esNonMyopic = nm
+          }
+      , e
+      ) <-
+    judgmentContext
+  let utxoSt = lsUTxOState ls
+  let DPState dstate pstate = lsDPState ls
+  ss' <-
+    trans @(EraRule "SNAP" era) $ TRC (ls, ss, ())
+
+  let PState pParams fPParams _ _ = pstate
+      ppp = eval (pParams ⨃ fPParams)
+      pstate' =
+        pstate
+          { psStakePoolParams = ppp
+          , psFutureStakePoolParams = Map.empty
+          }
+  PoolreapState utxoSt' acnt' dstate' pstate'' <-
+    trans @(EraRule "POOLREAP" era) $
+      TRC (pp, PoolreapState utxoSt acnt dstate pstate', e)
+
+  let adjustedDPstate = DPState dstate' pstate''
+      epochState' =
+        EpochState
+          acnt'
+          ss'
+          (ls {lsUTxOState = utxoSt', lsDPState = adjustedDPstate})
+          pr
+          pp
+          nm
+
+  let
+    Coin oblgNew = obligationDPState adjustedDPstate
+    Coin reserves = asReserves acnt'
+    utxoSt''' = utxoSt' {sutxosDeposited = Coin oblgNew}
+    acnt'' = acnt' {asReserves = Coin reserves}
```

Comment:
No need to unwrap the `Coin`, just to wrap it back up again. In fact those bindings are used only once, so inlining `reserves` makes it more readable:
```suggestion
    oblgNew = obligationDPState adjustedDPstate
    utxoSt''' = utxoSt' {sutxosDeposited = oblgNew}
    acnt'' = acnt' {asReserves = asReserves acnt'}
```

`oblgNew` is a bit more subjective, but that's the way I'd write it
```haskell
    utxoSt''' = utxoSt' {sutxosDeposited = obligationDPState adjustedDPstate}
    acnt'' = acnt' {asReserves = asReserves acnt'}
```